sale_coupon_product_management/models/sale_coupon_program.py

```diff
@@ -0,0 +1,179 @@
+# Copyright 2021 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, fields, models
+from odoo.exceptions import UserError
+
+
+class SaleCouponProgram(models.Model):
+    """Extend to improve product related product management."""
+
+    _inherit = "sale.coupon.program"
+
+    force_product_default_code = fields.Char()
+    # We would have "force_product_categ_id" required,
+    # but to not break all unit tests on coupon programs on other modules,
+    # we just add a required into the form view,
+    # and do nothing in this code if not set.
+    force_product_categ_id = fields.Many2one(
+        comodel_name="product.category", domain=[("is_program_category", "=", True)],
+    )
+    discount_line_product_chosen = fields.Boolean()
+    discount_line_product_readonly_id = fields.Many2one(
+        related="discount_line_product_id",
+    )
+
+    @api.onchange("program_type", "promo_applicability")
+    def _onchange_promo_applicability(self):
+        if (
+            self.program_type == "promotion_program"
+            and self.promo_applicability == "on_next_order"
+            and not self.force_product_categ_id
+        ):
+            # If several default categories found, just take the first
+            default_categ = self.env["product.category"].search(
+                [("default_promotion_next_order_category", "=", True)], limit=1
+            )
+            if default_categ:
+                self.force_product_categ_id = default_categ.id
+
+    @api.onchange("discount_line_product_chosen")
+    def _onchange_discount_line_product_chosen(self):
+        for rec in self:
+            rec.force_product_default_code = False
+            if rec.discount_line_product_id:
+                rec.discount_line_product_id = False
+
+    @api.onchange("force_product_categ_id")
+    def _onchange_force_product_categ_id(self):
+        for rec in self:
+            rec.discount_line_product_id = False
+
+    @api.constrains("force_product_categ_id", "reward_type", "discount_type")
+    def _check_program_options(self):
+        for rec in self:
+            category = rec.force_product_categ_id
+            if (
+                category
+                and category.program_product_discount_fixed_amount
+                and not (
+                    rec.reward_type == "discount"
+                    and rec.discount_type == "fixed_amount"
+                )
+            ):
+                raise UserError(
+                    _(
+                        "With 'program_product_discount_fixed_amount' category, "
+                        "the reward type must be 'discount' and "
+                        "the discount type must be 'Fixed Amount'."
+                    )
+                )
+
+    def _check_no_product_duplicate(self):
+        for rec in self:
+            other_program_found = self.search_count(
```

Comment:
can you search on recordset without loop

Reply:
No, I can't because I need to have the current program id to do the search.
The other solution can be to do it directly with SQL request, but I prefer to check it with python code.